// @segment/analytics-signals/dist/esm/lib/detect-url-change/index.js
import { Emitter } from "@segment/analytics-generic-utils";
// This seems hacky, but if using react router (or other SPAs), popstate will not always fire on navigation
// Otherwise, we could use popstate / hashchange events
export class URLChangeObservable {
    constructor(settings = {}) {
        this.emitter = new Emitter();
        this.pollInterval = settings.pollInterval ?? 500;
        this.pollURLChange();
    }
    pollURLChange() {
        let prevUrl = window.location.href;
        setInterval(() => {
            const currentUrl = window.location.href;
            if (currentUrl != prevUrl) {
                const current = new URL(currentUrl);
                const prev = new URL(prevUrl);
                this.emitter.emit("change", {
                    current,
                    previous: prev,
                });
                prevUrl = currentUrl;
            }
        }, this.pollInterval);
    }
    subscribe(urlChanged) {
        this.urlChanged = urlChanged;
        this.emitter.on("change", this.urlChanged);
    }
    unsubscribe() {
        this.emitter.off("change", this.urlChanged);
    }
}
