// @segment/analytics-signals/dist/cjs/core/signal-generators/dom-gen/navigation-gen.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OnNavigationEventGenerator = void 0;
const detect_url_change_1 = require("../../../lib/detect-url-change");
const factories_1 = require("../../../types/factories");
function getChangedProperties(url1, url2) {
    const changed = [];
    const propertiesToCompare = ["pathname", "search", "hash"];
    for (const property of propertiesToCompare) {
        if (url1[property] !== url2[property]) {
            if (property === "pathname") {
                changed.push("path");
            }
            else {
                changed.push(property);
            }
        }
    }
    return changed;
}
class OnNavigationEventGenerator {
    constructor() {
        this.id = "navigation";
        this.urlChange = new detect_url_change_1.URLChangeObservable();
    }
    register(emitter) {
        emitter.emit((0, factories_1.createNavigationSignal)({
            ...this.createCommonFields(),
        }));
        // emit a navigation signal whenever the URL has changed
        this.urlChange.subscribe(({ previous, current }) => emitter.emit((0, factories_1.createNavigationSignal)({
            previousUrl: previous.href,
            changedProperties: getChangedProperties(current, previous),
            ...this.createCommonFields(),
        })));
        return () => {
            this.urlChange.unsubscribe();
        };
    }
    createCommonFields() {
        return {
            // these fields are named after those from the page call, rather than a DOM api.
            currentUrl: location.href,
            path: location.pathname,
            hash: location.hash,
            search: location.search,
            title: document.title,
        };
    }
}
exports.OnNavigationEventGenerator = OnNavigationEventGenerator;
