// @segment/analytics-signals/dist/cjs/core/middleware/event-processor/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignalsEventProcessorSubscriber = void 0;
const logger_1 = require("../../../lib/logger");
const processor_1 = require("../../processor/processor");
const sandbox_1 = require("../../processor/sandbox");
class SignalsEventProcessorSubscriber {
    load(ctx) {
        this.buffer = ctx.buffer;
        const sandboxSettings = ctx.unstableGlobalSettings.sandbox;
        const normalizedEdgeFunctionURL = (0, sandbox_1.normalizeEdgeFunctionURL)(sandboxSettings.functionHost, sandboxSettings.edgeFnDownloadURL);
        let sandbox;
        if (!normalizedEdgeFunctionURL) {
            console.warn(`No processSignal function found. Have you written a processSignal function on app.segment.com?`);
            logger_1.logger.debug("Initializing sandbox: noop");
            sandbox = new sandbox_1.NoopSandbox();
        }
        else if (sandboxSettings.sandboxStrategy === "iframe" ||
            sandboxSettings.processSignal) {
            logger_1.logger.debug("Initializing sandbox: iframe");
            sandbox = new sandbox_1.WorkerSandbox(new sandbox_1.IframeSandboxSettings({
                processSignal: sandboxSettings.processSignal,
                edgeFnDownloadURL: normalizedEdgeFunctionURL,
            }));
        }
        else {
            logger_1.logger.debug("Initializing sandbox: global scope");
            sandbox = new sandbox_1.GlobalScopeSandbox({
                edgeFnDownloadURL: normalizedEdgeFunctionURL,
            });
        }
        this.processor = new processor_1.SignalEventProcessor(ctx.analyticsInstance, sandbox);
    }
    async process(signal) {
        return this.processor.process(signal, await this.buffer.getAll());
    }
}
exports.SignalsEventProcessorSubscriber = SignalsEventProcessorSubscriber;
